query-coordinator/src/main/scala/com/socrata/querycoordinator/QueryParser.scala

```diff
@@ -25,19 +27,57 @@ class QueryParser(analyzer: SoQLAnalyzer[SoQLAnalysisType], maxRows: Option[Int]
     }
   }
 
-  private def limitRows(analysis: SoQLAnalysis[ColumnName, SoQLAnalysisType])
-  : Either[Result, SoQLAnalysis[ColumnName, SoQLAnalysisType]] = {
-    analysis.limit match {
+  /**
+   * Convert analyses from column names to column ids.
+   * Add new columns and remap "column ids" as it walks the soql chain.
+   */
+  private def remapAnalyses(columnIdMapping: Map[ColumnName, String], //schema: Map[String, SoQLType],
+                            analyses: Seq[SoQLAnalysis[ColumnName, SoQLAnalysisType]])
+    : Seq[SoQLAnalysis[String, SoQLAnalysisType]] = {
+    val initialAcc = (columnIdMapping, Seq.empty[SoQLAnalysis[String, SoQLAnalysisType]])
+    val (_, analysesInColIds) = analyses.foldLeft(initialAcc) { (acc, analysis) =>
+      val (mapping, convertedAnalyses) = acc
+      // Newly introduced columns will be used as column id as is.
+      // There should be some sanitizer upstream that checks for field_name conformity.
+      // TODO: Alternatively, we may need to use internal column name map for new and temporary columns
+      val newlyIntroducedColumns = analysis.selection.keys.filter { columnName => !mapping.contains(columnName) }
+      val mappingWithNewColumns = newlyIntroducedColumns.foldLeft(mapping) { (acc, newColumn) =>
+        acc + (newColumn -> newColumn.name)
+      }
+      // Re-map columns except for the innermost soql
+      val newMapping =
+        if (convertedAnalyses.nonEmpty) {
+          val prevAnalysis = convertedAnalyses.last
+          prevAnalysis.selection.foldLeft(mapping) { (acc, selCol) =>
+            val (colName, expr) = selCol
+            acc + (colName -> colName.name)
+          }
+        } else {
+          mappingWithNewColumns
+        }
+
+      val a: SoQLAnalysis[String, SoQLAnalysisType] = analysis.mapColumnIds(newMapping)
+      (mappingWithNewColumns, convertedAnalyses :+ a)
+    }
+    analysesInColIds
+  }
+
+  private def limitRows(analyses: Seq[SoQLAnalysis[ColumnName, SoQLAnalysisType]])
+    : Either[Result, Seq[SoQLAnalysis[ColumnName, SoQLAnalysisType]]] = {
+    val lastAnalysis = analyses.last
+    lastAnalysis.limit match {
       case Some(lim) =>
         val actualMax = BigInt(maxRows.map(_.toLong).getOrElse(Long.MaxValue))
-        if (lim <= actualMax) Right(analysis) else Left(RowLimitExceeded(actualMax))
+        if (lim <= actualMax) { Right(analyses) }
+        else { Left(RowLimitExceeded(actualMax)) }
       case None =>
-        Right(analysis.copy(limit = Some(defaultRowsLimit)))
+        Right(analyses.dropRight(1) :+ lastAnalysis.copy(limit = Some(defaultRowsLimit)))
     }
   }
 
-  def apply(query: String, columnIdMapping: Map[ColumnName, String], schema: Map[String, SoQLType]): Result =
-    go(columnIdMapping, schema)(analyzer.analyzeUnchainedQuery(query)(_))
+  def apply(query: String, columnIdMapping: Map[ColumnName, String], schema: Map[String, SoQLType]): Result = {
+    go(columnIdMapping, schema)(analyzer.analyzeFullQuery(query)(_))
```

Comment:
It might be worth introducing a call to `SoQLAnalysis.merge` here; that will try to reduce the complexity of the parsed query by merging adjacent steps if possible.